src/db.rs

```diff
@@ -154,42 +192,52 @@ impl DB {
             }
         };
 
-        let sum_age = self.dirs.iter().map(|dir| dir.rank).sum::<Rank>();
+        let sum_age = self.data.dirs.iter().map(|dir| dir.rank).sum::<Rank>();
 
         if sum_age > max_age {
             let factor = 0.9 * max_age / sum_age;
-            for dir in &mut self.dirs {
+            for dir in &mut self.data.dirs {
                 dir.rank *= factor;
             }
 
-            self.dirs.retain(|dir| dir.rank >= 1.0);
+            self.data.dirs.retain(|dir| dir.rank >= 1.0);
         }
 
         self.modified = true;
         Ok(())
     }
 
     pub fn query(&mut self, keywords: &[String], now: Epoch) -> Option<Dir> {
-        let (idx, dir) = self
+        let (idx, dir, frecency) = self
```

Comment:
`frecency` appears to be unused. If you plan on using it later, should be fine to just prefix it with an underscore. Otherwise, it should probably be removed.

Reply:
Didn't see that, thanks!

Reply:
@ajeetdsouza (I think "resolved" comments no longer ping, so hoping the mention does)

Would it be possible to do something like this, instead? I really don't like seeing that unused 3rd tuple element when all it's necessary for is getting the max.

```rust
        let (idx, dir) = self
            .data
            .dirs
            .iter()
            .enumerate()
            .filter(|(_, dir)| dir.is_match(&keywords))
            .max_by(|(_, dir1), (_, dir2)| {
                let frecency1 = dir1.get_frecency(now);
                let frecency2 = dir2.get_frecency(now);

                frecency1.partial_cmp(frecency2).unwrap_or(Ordering::Equal)
            })?;
```

Reply:
@cole-h, while this did cross my mind, it results in 2 frecency calculations per `Dir` to the best of my knowledge, which is why I didn't go for it.

Reply:
@ajeetdsouza (ping again to ensure this gets to you)

I just did an extremely simple test on the playground to see the generated code.

Here's the source for the current version: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=dcbaf05896ecb011e2ef57fc7daf7e7b

And for my proposal: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7273d0c2d7d67796c0d9c0e77d4939c5

You can compare for yourself, but:

```shell
λ wc -l mir1 mir2
  1900 mir1 # proposed
  1951 mir2 # current
  3851 total
```

```shell
λ wc -l llvm1 llvm2
   9768 llvm1 # proposed
  10247 llvm2 # current
  20015 total
```

```shell
λ wc -l asm1 asm2
  4618 asm1 # proposed
  5181 asm2 # current
  9799 total
```

---

The above are from the debug builds, but I looked at the ASM release build and it was 424 (proposed) vs 464 (current).

Reply:
@cole-h, here's what I tried on my machine:
- create and add these directories into the database: `test1`, `test2`, `test3`, `test4`
- add a `println` or a `sleep` statement to `get_frecency`
- run a query: `zoxide query test`

You will notice that in the current case, there are 4 calls to `get_frecency`, one for each directory in the database. However, the proposed method has 6 calls (`test1` & `test2`, `test2` & `test3`, and so on). Other than the first and last matches, each directory makes 2 calls to `get_frecency`.

I too feel that the querying logic is quite a bit ugly, and I'd be open to figuring out ways of making it better, but making 2 calls to `get_frecency` per directory does seem wasteful.

Reply:
@ajeetdsouza 

Then we could do the following (ugly, but it's what I came up with a single minute of thought):

```rust
        let (idx, dir) = self
            .data
            .dirs
            .iter()
            .enumerate()
            .filter(|(_, dir)| dir.is_match(&keywords))
            .map(|(idx, dir)| (idx, dir, dir.get_frecency(now)))
            .max_by(|(_, _, frecency1), (_, _, frecency2)| {
                frecency1.partial_cmp(frecency2).unwrap_or(Ordering::Equal)
            })
            .map(|(idx, dir, _)| (idx, dir))?; // <---- new 
```

I'll do more thinking on this, but wanted to put it here because 2 minds are better than 1.

Reply:
@cole-h I don't see the benefit of the additional `map` call, since it seems to be more boilerplate than the original code. I think it's fine to use destructuring to ignore part of a tuple; it's a fairly common occurrence in Rust code. 

I don't think there's a way of simplifying the iterator itself, so I think we should leave this for the time being and approach it later on if a solution presents itself.

Reply:
Fair enough. Thanks for bikeshedding with me.